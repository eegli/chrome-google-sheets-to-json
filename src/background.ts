chrome.runtime.onInstalled.addListener(() => {
  // Page actions are disabled by default and enabled on select tabs
  //@ts-ignore the typings seem to be very alpha stagey
  chrome.action.disable();

  // Clear all rules to ensure only our expected rules are set
  chrome.declarativeContent.onPageChanged.removeRules(undefined, () => {
    // Declare a rule to enable the action on example.com pages
    let exampleRule = {
      conditions: [
        new chrome.declarativeContent.PageStateMatcher({
          pageUrl: { hostEquals: 'docs.google.com' },
        }),
        new chrome.declarativeContent.PageStateMatcher({
          pageUrl: { pathContains: 'spreadsheets/d' },
        }),
      ],
      actions: [new chrome.declarativeContent.ShowPageAction()],
    };

    // Finally, apply our new array of rules
    let rules = [exampleRule];
    chrome.declarativeContent.onPageChanged.addRules(rules);
  });
});
