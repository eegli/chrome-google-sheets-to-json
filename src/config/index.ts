export default {
  defaultSheet: '1',
};
