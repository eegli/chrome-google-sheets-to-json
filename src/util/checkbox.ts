export function insertCheckboxes(parent: HTMLElement, data: string[]): void {
  let idx = data.length - 1;
  // Looping backwards through the sheet names and appending a new node before the first child of the parent

  for (; idx >= 0; idx--) {
    const el = data[idx];
    const container = document.createElement('div');
    const input = document.createElement('input');
    const label = document.createElement('label');

    input.type = 'radio';
    input.name = 'sheetSelect';
    input.value = el;
    input.id = el;
    if (idx === 0) input.checked = true;

    label.htmlFor = el;
    label.innerHTML = el;

    container.classList.add('form-check');
    input.classList.add('form-check-input');
    label.classList.add('form-check-label');

    container.append(input);
    container.append(label);

    parent.prepend(container);
  }
}
