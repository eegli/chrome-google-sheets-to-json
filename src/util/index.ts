import * as GST from 'google-spreadsheets-ts';

type DownloadArgs = {
  url: string;
  page: number;
  fileName: string;
};

export async function downloadJSON(opts: DownloadArgs): Promise<void> {
  const { url, page, fileName } = opts;
  const JSONendpoint = getJSONEndpoint(url, page);
  try {
    // Fetch data, extract and prompt download
    const response = await fetch(JSONendpoint);
    const data: GST.RootObject = await response.json();
    const json = extractJSON(data);
    const jsonStr = JSON.stringify(json);
    const blob = new Blob([jsonStr], { type: 'application/json' });
    const downloadUrl = URL.createObjectURL(blob);

    chrome.downloads.download({
      url: downloadUrl,
      filename: fileName + '.json'
    });
  } catch (e) {
    throw new Error('Error downloading JSON');
  }
}

function getJSONEndpoint(docsUrl: string, page: number): string {
  const [sheetId] = docsUrl.match('(?<=\\/d\\/)[^\\/]*') || [''];
  return `https://spreadsheets.google.com/feeds/list/${sheetId}/${page}/public/values?alt=json`;
}

function extractJSON(data: GST.RootObject): any[] {
  const res: Record<string, any>[] = [];

  const rows = data.feed.entry;

  // Empty Google Sheet
  if (!rows) return res;

  for (const row of rows) {
    const formattedRow = {} as Record<string, any>;

    for (const key in row) {
      if (key.startsWith('gsx$') && !key.endsWith('_cpzh4')) {
        formattedRow[key.replace('gsx$', '')] = row[key].$t;
      }
    }

    res.push(formattedRow);
  }

  return res;
}
