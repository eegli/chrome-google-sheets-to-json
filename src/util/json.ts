import * as GST from 'google-spreadsheets-ts';

export function getJSONEndpoint(id: string, page: number): string {
  return `https://spreadsheets.google.com/feeds/list/${id}/${page}/public/values?alt=json`;
}
export function extractJSON(data: GST.RootObject): any[] {
  const res: Record<string, any>[] = [];

  const rows = data.feed.entry;

  // Empty Google Sheet
  if (!rows) return res;

  for (const row of rows) {
    const formattedRow = {} as Record<string, any>;

    for (const key in row) {
      if (key.startsWith('gsx$') && !key.endsWith('_cpzh4')) {
        formattedRow[key.replace('gsx$', '')] = row[key].$t;
      }
    }

    res.push(formattedRow);
  }

  return res;
}
