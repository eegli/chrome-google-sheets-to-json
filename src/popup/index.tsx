import { FunctionalComponent, render } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import { sheetNamesContentScript } from '../scripts/sheet';
import { downloadJSON } from '../util';
import CheckBoxes from './checkbox';

export type Sheets = string[];

const Popup: FunctionalComponent = () => {
  const [activeTab, setActiveTab] = useState<chrome.tabs.Tab>();
  const [sheets, setSheets] = useState<Sheets>([]);
  const [selectedSheet, setSelectedSheet] = useState<string>('');
  const [isPublicDoc, setIsPublicDoc] = useState<boolean>(true);

  /* https://preactjs.com/guide/v10/typescript/#typing-events */
  function handleSelectSheet(event: Event) {
    event.preventDefault();
    if (event.target instanceof HTMLInputElement) {
      setSelectedSheet(event.target.value);
    }
  }

  function handleSubmit(event: Event): void {
    event.preventDefault();
    if (event.target instanceof HTMLFormElement) {
      handleDownload();
    }
  }

  async function handleDownload(): Promise<void> {
    // Each sheet has its own url. The first sheet has number 1, the
    // second number 2, etc
    const page = sheets.indexOf(selectedSheet) + 1;
    const url = activeTab?.url || '';
    const fileName =
      activeTab?.title?.replace(' - Google Sheets', '') || 'data';

    // Download JSON
    try {
      await downloadJSON({ url, page, fileName });
      window.close();
    } catch (e) {
      console.log('Sheet is not public');
      console.error('An error occured', e);
      setIsPublicDoc(false);
    }
  }

  // https://dev.to/stlnick/useeffect-and-async-4da8
  useEffect(() => {
    (async () => {
      const [tab] = await chrome.tabs.query({
        active: true,
        currentWindow: true
      });

      setActiveTab(tab);

      const sheets = await chrome.scripting.executeScript({
        target: { tabId: tab.id || 0 },
        func: sheetNamesContentScript
      });

      const sheetNames: string[] = sheets[0].result;

      setSheets(sheetNames);
      setSelectedSheet(sheetNames[0]);
    })();
  }, []);

  return (
    <main>
      <div className="container w-auto p-3">
        {isPublicDoc ? (
          <div>
            <h4 className="text-nowrap">Choose Sheet</h4>
            <form name="sheet" onSubmit={handleSubmit}>
              <CheckBoxes
                sheets={sheets}
                selectedSheet={selectedSheet}
                onChange={handleSelectSheet}
              />
              <button type="submit" className="btn btn-primary my-3">
                Submit
              </button>
            </form>
          </div>
        ) : (
          <div className="alert alert-danger" role="alert">
            <h4 className="alert-heading text-nowrap">Sheet is not public</h4>
            <p>
              It looks like the Sheet "{selectedSheet}" is not published. As of
              now, this extension can only download Google Sheets that are
              published to the web.
            </p>
            <hr />
            <p className="mb-0">
              Publish your Sheet and try again.
              <br />
              <a
                href="https://support.google.com/docs/answer/183965"
                rel="noopener noreferrer"
                target="_blank"
                class="alert-link">
                How to publish a Google Sheet
              </a>
            </p>
          </div>
        )}
      </div>
    </main>
  );
};

render(<Popup />, document.body);
