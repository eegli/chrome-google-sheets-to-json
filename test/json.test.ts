/**
 * @jest-environment jsdom
 */

import googleJSON from './data/sample-spreadsheet';
import { download, extractJSON, getEndpoint, fetchDocsData } from '../src/util';

const mockedRes = { hello: 'jest' };

const mockedFetch = jest.fn().mockResolvedValueOnce({
  json: async () => Promise.resolve(mockedRes)
} as Response);

window.fetch = mockedFetch;

describe('Fetching Docs data', () => {
  it('fetches Google Docs data', async () => {
    const res = await fetchDocsData('test');
    expect(mockedFetch).toHaveBeenCalledTimes(1);
    expect(res).toEqual(mockedRes);
  });
});

describe('JSON parsing and utilities', () => {
  it('generates JSON from the raw response', () => {
    const json = extractJSON(googleJSON);
    expect(json).toMatchSnapshot();
  });

  it('extracts Sheet id and returns an url', () => {
    const urls = [
      'https://docs.google.com/spreadsheets/d/' +
        '1qkr79b8V3GR2Uk2ZYCzeccm7oWWeNV72TZRKOL_Khbg' +
        '/edit',
      'https://docs.google.com/spreadsheets/d/' +
        '1M3qWk9Qvh87qY3pdMB53RwQ513QcUtps9t71I-S-pUk',
      'https://docs.google.com/spreadsheets/d/' +
        '11CVwVoQyn9a2AgztdDf-EZHS0aHxFiHoZu_aFr1PDXk' +
        '/edit#gid=771262429'
    ];
    const mockedGetEndpoint = jest.fn(getEndpoint);
    urls.forEach(url => mockedGetEndpoint(url, 1));

    expect(mockedGetEndpoint.mock.results[0].value).toEqual(
      'https://spreadsheets.google.com/feeds/list/' +
        '1qkr79b8V3GR2Uk2ZYCzeccm7oWWeNV72TZRKOL_Khbg' +
        '/' +
        '1' +
        '/public/values?alt=json'
    );
    expect(mockedGetEndpoint.mock.results[1].value).toEqual(
      'https://spreadsheets.google.com/feeds/list/' +
        '1M3qWk9Qvh87qY3pdMB53RwQ513QcUtps9t71I-S-pUk' +
        '/' +
        '1' +
        '/public/values?alt=json'
    );
    expect(mockedGetEndpoint.mock.results[2].value).toEqual(
      'https://spreadsheets.google.com/feeds/list/' +
        '11CVwVoQyn9a2AgztdDf-EZHS0aHxFiHoZu_aFr1PDXk' +
        '/' +
        '1' +
        '/public/values?alt=json'
    );
  });

  it('prompts a download', () => {
    const downloadSpy = jest.fn();
    window.URL.createObjectURL = jest.fn();
    chrome.downloads.download = jest.fn(downloadSpy);

    download({ fileName: 'test', data: {} });

    expect(downloadSpy).toHaveBeenCalledTimes(1);
  });
});
