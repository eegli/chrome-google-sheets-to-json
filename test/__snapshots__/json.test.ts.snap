// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON parsing and utilities generates JSON from the raw response 1`] = `
Array [
  Object {
    "age": "14",
    "firstname": "Tinu",
    "lastname": "Elejogun",
  },
  Object {
    "age": "28",
    "firstname": "Javier",
    "lastname": "Zapata",
  },
  Object {
    "age": "18",
    "firstname": "Lily",
    "lastname": "McGarrett",
  },
  Object {
    "age": "22",
    "firstname": "Olatunkbo",
    "lastname": "Chijiaku",
  },
  Object {
    "age": "22",
    "firstname": "Adrienne",
    "lastname": "Anthoula",
  },
  Object {
    "age": "22",
    "firstname": "Axelia",
    "lastname": "Athanasios",
  },
  Object {
    "age": "22",
    "firstname": "Jon-Kabat",
    "lastname": "Zinn",
  },
  Object {
    "age": "15",
    "firstname": "Thabang",
    "lastname": "Mosoa",
  },
  Object {
    "age": "11",
    "firstname": "Kgaogelo",
    "lastname": "Mosoa",
  },
]
`;
